src/webob/acceptparse.py

```diff
@@ -255,11 +294,371 @@ class AcceptEncoding(Accept):
+    docstring for :meth:`AcceptLanguageValidHeader.__add__`).
+    """
+
+    # RFC 7231 Section 5.3.5 "Accept-Language":
+    # Accept-Language = 1#( language-range [ weight ] )
+    # language-range  =
+    #           <language-range, see [RFC4647], Section 2.1>
+    # RFC 4647 Section 2.1 "Basic Language Range":
+    # language-range   = (1*8ALPHA *("-" 1*8alphanum)) / "*"
+    # alphanum         = ALPHA / DIGIT
+    lang_range_re = r"""
+        \*|
+        (?:
+        [A-Za-z]{1,8}
+        (?:-[A-Za-z0-9]{1,8})*
+        )
+    """
+    lang_range_n_weight_re = _item_n_weight_re(item_re=lang_range_re)
+    lang_range_n_weight_compiled_re = re.compile(
+        lang_range_n_weight_re,
+        re.VERBOSE
+    )
+    accept_language_compiled_re = _list_1_or_more__compiled_re(
+        element_re=lang_range_n_weight_re,
+    )
+
+    def __init__(self, header_value):
+        """
+        Create an :class:`AcceptLanguageValidHeader` instance.
+
+        :param header_value: (``str``) header value.
+        :raises ValueError: if `header_value` is an invalid value for an
+                            ``Accept-Language`` header.
+        """
+        self._header_value = header_value
+        self._parsed = list(self.parse(header_value))
+        self._parsed_nonzero = [(m, q) for (m, q) in self.parsed if q]
+
+    @property
+    def header_value(self):
+        """(``str``) The header value."""
+        return self._header_value
+
+    @property
+    def parsed(self):
+        """
+        (``list``) Parsed form of the header.
+
+        A list of (language range, quality value) tuples.
+        """
+        return self._parsed
+
+    @classmethod
+    def parse(cls, value):
+        """
+        Parse an ``Accept-Language`` header.
+
+        :param value: (``str``) header value
+        :return: If `value` is a valid ``Accept-Language`` header, returns a
+                 generator that yields (language range, quality value) tuples,
+                 as parsed from the header from left to right.
+        :raises ValueError: if `value` is an invalid header
+        """
+        # Check if header is valid
+        # Using Python stdlib's `re` module, there is currently no way to check
+        # the match *and* get all the groups using the same regex, so we have
+        # to use one regex to check the match, and another to get the groups.
+        if cls.accept_language_compiled_re.match(value) is None:
+            raise ValueError  # invalid header
+        else:
+            def generator(value):
+                for match in (
+                    cls.lang_range_n_weight_compiled_re.finditer(value)
+                ):
+                    lang_range = match.group(1)
+                    qvalue = match.group(2)
+                    qvalue = float(qvalue) if qvalue else 1.0
+                    yield (lang_range, qvalue)
+            return generator(value=value)
```

Comment:
It is so that the `ValueError` is raised when `.parse()` is called, rather than when `.next()`/`.__next__()` is first called on the iterator. So if it's not a closure, we would have to attempt to consume the first item from the iterator before we know whether the header is invalid. But it could work either way, and depends on when you want the `ValueError` to be raised.

(The `else:` is unnecessary &mdash; I'll remove it.)